Windows Migration: Decoupling & Refactoring Extension Logger, to implement new interface while using slog.

klugorosado opened the pull request:
> [!IMPORTANT]
> ### Summary
> 1. Decoupled Logger from main package.
> 2. Created New Logger interface.
> 3. Implemented Extension Logger as an adapter for log/slog.
> 4. Upgraded Codebase from Go 1.18 to Go 1.21
> 5. Added New UTs for Logger and Bats integration Tests to use Bats-core add-on libraries like bats-assert.
> 6. Ran Go mod tidy && Go mod vendor.

## Feature: Application Health Extension Windows Migration: Subtask: Decoupling & Refactoring Extension Logger

### Description:
This feature involves migrating the codebase from.NET Framework to Golang to enable TLS 1.3 for Application Health Extension Windows. The Linux codebase will be made cross-platform. 


### This Pull Request: Decoupling & Refactoring Extension Logger. 
This pull request includes several updates to the project’s development environment, dependencies, and logging system. The changes primarily involve the addition of new dependencies and tools in the .devcontainer setup, updates to the Go dependencies in go.mod, and a transition from the go-kit/kit/log package to a custom logging package throughout various Go files and used the new go-kit/log package. 

### Important Changes:
This pull request introduces several significant changes to the development environment, extension settings, and the application's internal structure. The most notable changes include the addition of new dependencies and tools in the Dockerfile, updates to the `.devcontainer/devcontainer.json` and `.vscode/extensions.json` files to improve development experience, and a major refactor of the Go codebase to improve the structure and readability of the code.

### Related Pull Requests:
None
### Testing:

- Unit tests and integration tests have been conducted to ensure the stability of the new layout. 

### Milestones and Upcoming PR:
```[tasklist]
### Tasks
- [x] Decouple Logging from main package.
- [ ] Decouple Command Handlers from main package
- [ ] Decouple Plugin Code (AppHealth and VMWatch) from main package
- [ ] Create Windows Command Handler including pre functions.
- [ ] Decouple Resource Governance from VMWatch & add Windows Resource Governance. 
- [ ] Make applicationhealth-shim work with windows or find another way to start windows extension. 
- [ ] Add Windows build commands to Makefile
- [ ] Make Integration Tests work with Windows. 
```

dpoole73 commented on vendor/github.com/Azure/azure-extension-platform/pkg/utils/utils_windows.go:
```diff
@@ -0,0 +1,13 @@
+
+func GetDataFolder(name string, version string) string {
+	systemDriveFolder := os.Getenv("SystemDrive")
+	return path.Join(systemDriveFolder, "Packages\\Plugins", name, version, "Downloads")
```
this looks like windows not linux...

dpoole73 commented on main/vmWatch.go:
```diff
@@ -128,16 +128,15 @@ func executeVMWatchHelper(ctx *log.Context, attempt int, vmWatchSettings *vmWatc
-	ctx.Log("event", fmt.Sprintf("Attempt %d: VMWatch process started with pid %d", attempt, pid))
-
+	ctx.Event(fmt.Sprintf("Attempt %d: VMWatch process started with pid %d", attempt, pid))
 	err = createAndAssignCgroups(ctx, pid)
```
i don't see the windows code paths here, cgroups don't exist in windows, you'll need to port all that Win32Api stuff for job objects etc over here and depending on the OS you'll need to call different code paths

klugorosado replied:
This PR is the first of many to migrate windows extension. This PR focuses on decoupling the Logging. However I will keep this in mind since it will come later.

dpoole73 commented on integration-test/run.sh:
```diff
@@ -21,10 +21,10 @@ done
 create_certificate
 # Run Sequential Integration Tests
-sudo bats integration-test/test/sequential -T --trace 
+bats integration-test/test/sequential -T --trace 
```
how are these tests going to validate windows code paths?  seems like they will only run in linux environment right?

klugorosado replied:
Good Questions, Windows will have to differ from Linux.

dpoole73 commented on main/vmWatch.go:
```diff
@@ -363,11 +362,11 @@ func GetProcessDirectory() (string, error) {
 
-func GetExecutionEnvironment(hEnv HandlerEnvironment) string {
+func GetExecutionEnvironment(hEnv handlerenv.HandlerEnvironment) string {
 	if strings.Contains(hEnv.HandlerEnvironment.LogFolder, AppHealthPublisherNameTest) {
```
paths are different in windows and linux right?  i think this logic may need checking as it might not work the same between windows and linux

klugorosado replied:
I will address this on a later PR, I will decouple and separate the generic extension code, the plugin code (AppHealth, VMWatch) and the GHP Code (Resource Governance, heartbeat etc.)

dpoole73 commented on main/vmWatch.go:
```diff
@@ -128,16 +128,15 @@ func executeVMWatchHelper(ctx *log.Context, attempt int, vmWatchSettings *vmWatc
-	ctx.Log("event", fmt.Sprintf("Attempt %d: VMWatch process started with pid %d", attempt, pid))
-
+	ctx.Event(fmt.Sprintf("Attempt %d: VMWatch process started with pid %d", attempt, pid))
 	err = createAndAssignCgroups(ctx, pid)
```
this won't work on windows.  you will need to port all the job object stuff over here

dpoole73 commented on main/vmWatch.go:
```diff
@@ -203,22 +202,22 @@ func monitorHeartBeat(ctx *log.Context, heartBeatFile string, processDone chan b
 }
 
-func setupVMWatchCommand(s *vmWatchSettings, hEnv HandlerEnvironment) (*exec.Cmd, error) {
+func setupVMWatchCommand(s *vmWatchSettings, hEnv handlerenv.HandlerEnvironment) (*exec.Cmd, error) {
```
I don't see the changes here to deal with thre different binary names on windows/linux.  here are some constants there that specify the binary name but they are linux names you'll need to add code to those functions to get the correct name

klugorosado replied:
once the code is decoupled, I will add this type of function in a _windows or _linux.go file.

frank-pang-msft commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	if logFileFormat == "" {
+		logFileFormat = "log_%v"
```
Looks like log file format is `log_unixtimestamp`, does this differ from the existing naming convention?

klugorosado replied:
WHat do you mean exactly? For Linux we dont log to a file, For windows it Seems like we use AppHealthExtension.log

frank-pang-msft replied:
We do log to a file for Linux. Extension execution output is written to this folder.

/var/log/azure/applicationhealth-extension

frank-pang-msft replied:
I think it should be /var/log/azure/applicationhealth-extension/handler.log

klugorosado replied:
The logger has a multi writer, it will write to both "log_%v" files inside the LogFolder and to the Standard output.

frank-pang-msft replied:
Got it, so in the future when everything is supported. This is where our telemetry will go

1. Std output, and will also go to handler.log
2. LogFolder specified by HandlerEnvironment
3. GuestAgent EventsFolder

frank-pang-msft commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,226 @@
+func rotateLogFolder(logFolder string, logFileFormat string) (err error) {
+	size, err := getDirSize(logFolder)
+	if err != nil {
+		return
```
should this return the error?

frank-pang-msft replied:
same for below return statements.

is explicit return of err variable not required since return variable is named the same?

klugorosado replied:
this is returning an error, it will return the err variable defined in (err error)

frank-pang-msft commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,226 @@
+
+// Allows the caller to specify their own name for the file
+// Supports cycling of logs to prevent filling up the disk
+func NewWithName(he *handlerenv.HandlerEnvironment, logFileFormat string) *ExtensionLogger {
```
Thoughts on making UTs for this class? 

I think it might be hard to test the log rotation, but it might be good to test that prefixes are in the desired order?

frank-pang-msft replied:
Not sure how easy it will be to validate the ordering actually, so as long as you've checked in the existing integration tests that the log format is the desired outcome and doesn't differ as much from the existing convention, I'm good with that.

klugorosado replied:
ack

klugorosado replied:
Created new UTs to validate rotateLogFolder.

frank-pang-msft reviewed (changes requested):
Left some comments. logging should still work so as long as the integration tests still pass we should be fine. Just make sure to validate the formatting of the logs is expected.

Do you know how many PRs you'll be doing as a part of windows migration? Might be good to include each milestone somewhere so we can see when this feature is done.

klugorosado commented on .vscode/extensions.json:
```diff
@@ -1,7 +1,10 @@
+        "ms-vscode-remote.remote-ssh",
+        "github.copilot",
+        "ms-azuretools.vscode-docker",
+        "mhutchie.git-graph",
```
Not really, I put them here for them to show up as recommended when dev container is newly built.

klugorosado commented on .devcontainer/devcontainer.json:
```diff
@@ -16,22 +16,25 @@
             "extensions": [
-                "golang.go"
+                "golang.go",
+                "golang.go-nightly",
+                "premparihar.gotestexplorer",
```
go-nightly is, but will remove premparihar.gotestexplorer

frank-pang-msft commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+func (lg *ExtensionLogger) Event(event string) {
+	level.Info(lg.logger).Log(logEvent, event)
```
In Windows the pattern is to have a single logger that exposes methods writing to file vs to kusto, my preference is that, but I think what you suggested for run command will work too.

Just want to make sure this whole package will be easily refactorable and not thrown away

Hafthor commented on .devcontainer/devcontainer.json:
```diff
@@ -16,22 +16,24 @@
             "extensions": [
-                "golang.go"
+                "golang.go",
+                "golang.go-nightly",
```
curious what is making this a requirement. How soon can we just depend on the normal release channel version of Go?

klugorosado replied:
I will move "golang.go-nightly" to recommendations, however "golang.go" is needed to develop with go in vs code, and docker for the integration tests since they create new containers.

Hafthor commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+// Close closes the file
+func (logger *ExtensionLogger) Close() {
```
question: should this return `error` - I would imagine that logger.file.Close() does.

Hafthor commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,226 @@
+			size += info.Size()
+		}
+
+		return err
```
nit: I would `return nil;` to make it clear that this is the err==nil path.

klugorosado commented:
### New Changes:

1. Updates codebase from Go 1.18 -> Go 1.21.
2. Changed Logger from go-kit/log to Standards library slog/log.
3. Created Extension Logger Interface to allow use to swap out Logger in the future if needed.
4. Added New Unit Test for Logger. Tested the log file rotation and Format.

Upcoming Changes on Another PR:
1. Adding Adapter for Logger to use Extension-Platform Event Telemetry package.
2. Decoupling Extension & Plugin logics from main package.

klugorosado commented:
### New Changes:

1. Updates codebase from Go 1.18 -> Go 1.21.
2. Changed Logger from go-kit/log to Standards library slog/log.
3. Created Extension Logger Interface to allow use to swap out Logger in the future if needed.
4. Added New Unit Test for Logger. Tested the log file rotation and Format.
5. GitHub Workflow, Now runs with GO 1.21.
6. Added Bats add-on libraries to facility REGEX testing. They are installed automatically on GitHub Workflow and Dev Container.

### Upcoming Changes on Another PR:
1. Adding Adapter for Logger to use Extension-Platform Event Telemetry package.
2. Decoupling Extension & Plugin logics from main package.

frank-pang-msft commented on .devcontainer/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM mcr.microsoft.com/devcontainers/go:1.18-bullseye
+FROM mcr.microsoft.com/devcontainers/go:1.21-bullseye
```
It's best we keep this in sync with what's on LinuxExtensions repo, since that's where our code is built.

Did they update to 1.21 or did you need dependencies from go 1.21?

frank-pang-msft replied:
Ahh is it because log/slog package is not available until 1.21?

klugorosado replied:
There was an ask a few weeks ago to upgrade all Golang VM Extension to the latest Golang version because there was a vulnerability found. Thats why I am doing this, now

klugorosado replied:
Yes, log/slog becomes part of the standard lib on GO 1.21 + but there are older slog version from before it became the standard.

frank-pang-msft commented on .github/workflows/go.yml:
```diff
@@ -48,10 +49,26 @@ jobs:
       run: |
         sudo apt install npm
         sudo npm install -g bats
+        
```
nit spacing

klugorosado replied:
ack

frank-pang-msft commented on .github/workflows/go.yml:
```diff
@@ -48,10 +49,26 @@ jobs:
         sudo npm install -g bats
+        
+
+    - name: Setup bats libs
```
Haven't used this before, what's it do?

klugorosado replied:
bats-assert provide really nice helper function for bats test, For example instead of having [[ $output == *"something"*]] which it parses the entire log gile we can use: assert_line() to validate a specific line or vvalidate a regex on a specific line. 

I used it to validated the new logging format where there was non-determinitic values like pid in between values that I wanted to valided like msg, operation and path. you can see it in 7_vmwatch.bats

frank-pang-msft commented on integration-test/test/parallel/7_vmwatch.bats:
```diff
@@ -1,8 +1,8 @@
-load ../test_helper
-
 setup(){
+    load "../test_helper"
```
what issue did you run into? how come loading the script is required prior to each test?

klugorosado replied:
Ran into many problems, I ended doing this because loading the helper libaries was failing, due to inconsistent ways of them being installed, on the dev container I use npm, while on the Github Workflow I use the bats-action, why?

while npm was working on dev container, It depended on an enviroment variable since paths could change, I tried the same with the Githbub workflow, but the way it handles enviroment variables is not documented well, and Enviroment variables never reached the bats files.

Furthermore, if using bats (from bats-action) sudo couldn't be used which is needed to run in Parallel.

So Essentially I ended up using npm for both bats and libs on dev container, and bats via npm, libs via bats-action for github action. Then created the _load_bats_libs function to dynamicly select the correct path for the libraries.

frank-pang-msft commented on integration-test/test/test_helper.bash:
```diff
@@ -6,20 +6,22 @@ TEST_CONTAINER=test
-# If the image already exists, a random number is appended to the name.
-# a unique name is needed to avoid conflicts with other tests while running in parallel.
+_load_bats_libs() {
+    export BATS_LIB_PATH=${CUSTOM_BATS_LIB_PATH:-"/usr/lib:/usr/local/lib/node_modules"}
```
Is this dependencies required for devcontainer and not for local testing? Just wondering how we didn't run into issues before.

klugorosado replied:
What you mean by local testing? I do my local testing on the dev container, since it has access to docker.

This dependency is only needed now, where a complex regex is made on 7_vmwatch.bats.

frank-pang-msft commented on integration-test/test/test_helper.bash:
```diff
@@ -191,19 +192,20 @@ copy_config() { # places specified settings file ($1) into container as 0.settin
 verify_state_change_timestamps() {
     expectedTimeDifferences="$2"
-    regex='time=(.*) version=(.*)'
+    regex='time=([^[:space:]]*)'
```
what's the format look like now? Might be good to update as a comment above.

klugorosado replied:
Same format as before, the only difference it will not be looking for version to get the time, it will just stop when it finds the first space.

Will update the doc

frank-pang-msft commented on integration-test/test/test_helper.bash:
```diff
@@ -213,7 +215,7 @@ verify_state_change_timestamps() {
 verify_states() {
     expectedStateLogs="$2"
-    regex='event="(.*)"'
+    regex='msg="(.*)"'
```
wondering what your thoughts were about changing event vs msg?

frank-pang-msft replied:
Ahh nvm, i think your new logging changes make it so there is log.Level and log.msg, you can disregard other comment about ctx.Info as well

klugorosado replied:
ack

frank-pang-msft commented on main/cmds.go:
```diff
@@ -39,31 +41,32 @@ var (
 	{ // a new context scope with path
-		ctx = ctx.With("path", dataDir)
-		ctx.Log("event", "removing data dir", "path", dataDir)
+		ctx.With("path", dataDir)
```
just confirming we don't need reassignment of ctx right?
e.g
ctx = ctx.With("path", dataDir)

klugorosado replied:
no, that is done internally inside the Logger.

frank-pang-msft commented on main/cmds.go:
```diff
@@ -39,31 +41,32 @@ var (
-func noop(ctx *log.Context, h HandlerEnvironment, seqNum int) (string, error) {
-	ctx.Log("event", "noop")
+func noop(ctx logging.Logger, h handlerenv.HandlerEnvironment, seqNum int) (string, error) {
+	ctx.Info("noop")
```
so ctx.Info() is replacing ctx.Log("event"), is there reason why we do "msg" prefix instead of "info"?

frank-pang-msft replied:
nvm, it's cause you have log levels as well

klugorosado replied:
msg prefix is the standard in slog, it has 3 default fields time, level, msg, however we can change the Keys for each of them, for example we could have UTCTime, level, event.

or ww could create a grouping like time, level, msg, event.message=, event.type=, event.details=

frank-pang-msft replied:
i'm okay with the current formatting

frank-pang-msft commented on internal/handlerenv/handlerenv.go:
```diff
@@ -1,4 +1,4 @@
-package main
+package handlerenv
```
Is there plan to leverage azure-extension-platform handlerenv eventually?

klugorosado replied:
Yes we can, I think I can look more into it next PR since I moved the version.go to another pkg, I have to re-read the documentation carefully, to make sure what Version they need to what file for Linux.

Also we hardcode the Version in status.go we might want to change that behavior as well.

frank-pang-msft commented on main/cmds.go:
```diff
@@ -219,7 +223,7 @@ func enable(ctx *log.Context, h HandlerEnvironment, seqNum int) (string, error)
 		err = reportStatusWithSubstatuses(ctx, h, seqNum, StatusSuccess, "enable", statusMessage, substatuses)
 		if err != nil {
-			ctx.Log("error", err)
+			ctx.Error("Failed to report status", slog.Any("error", err))
```
Just getting myself familiar with the format. 

Does `slog.Any("error", err)` make a json key value pair similar to {"error": myErrorHere}?

klugorosado replied:
Yes, it's similar, it created this specifically will create a log with the following format `time={time} level=error msg="Failed to report status" error={myErrorHere}`

But since I added custom formatting, it will format the error key into a custom group and change the format to the following: `level=error msg="Failed to report status" error.message= {myErrorHere message}, error.type= {myErrorHere Type} error.stacktrace={stacktrace} `

frank-pang-msft commented on go.mod:
```diff
@@ -1,13 +1,14 @@
 
 require (
 	github.com/Azure/azure-docker-extension v0.0.0-20160802215703-0dd2f199467d
+	github.com/Azure/azure-extension-platform v0.0.0-20230218002700-ca684482c954
```
What steps did you follow to update go.mod, go.sum and vendor files?

Was it `go get -u` and `go mod vendor` or `go mod tidy`?

klugorosado replied:
I always do go mod tidy -> go mod vendor.

dpoole73 commented on main/vmWatch.go:
```diff
@@ -93,32 +95,32 @@ func executeVMWatch(ctx *log.Context, s *vmWatchSettings, hEnv HandlerEnvironmen
 		}
-		ctx.Log("error", fmt.Sprintf("VMWatch reached max %d retries, sleeping for %v hours before trying again", VMWatchMaxProcessAttempts, HoursBetweenRetryAttempts))
+		err := fmt.Errorf("VMWatch reached max %d retries, sleeping for %v hours before trying again", VMWatchMaxProcessAttempts, HoursBetweenRetryAttempts)
+		ctx.Error("VMWatch reached max retries", slog.Any("error", err))
```
could you explain what this `slog.Any()` stuff does?

klugorosado replied:
New Logger is based Go 1.21+ standard logger "log/slog". With the slog logger we can group keys together.

For the Error logs its needed, because the logger finds a key named error, it will add additional formatting to the log, for example it will add error.message, error.type and error.stacktrace.

dpoole73 commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,266 @@
+)
+
+const (
+	thirtyMB            = 30 * 1024 * 1034 // 31,457,280 bytes
```
1024?

klugorosado replied:
nice catch, Thanks!

dpoole73 commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,266 @@
+// a standard output logger will be used.
+// It returns a Logger instance.
+func NewExtensionLogger(he *handlerenv.HandlerEnvironment) Logger {
+	if he == nil || he.HandlerEnvironment.LogFolder == "" {
```
can `he.HandlerEnvironment` also be nil? do we need to check?

klugorosado replied:
I dont think he.HandlerEnviroment can be nil because its embedded.

dpoole73 commented on pkg/logging/logging.go:
```diff
@@ -0,0 +1,266 @@
+	fileName := fmt.Sprintf(logFileFormat, strconv.FormatInt(time.Now().UTC().Unix(), 10))
+	filePath := path.Join(logFolder, fileName)
+	writer, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
+	if err != nil {
```
this feels weird to me, eating the error when failing to create the file.  Shouldn't you just fail here (ie you've been asked to log to a file and failed to do so, feels like a failure case to me)

klugorosado replied:
ack

dpoole73 commented on main/cmds.go:
```diff
@@ -39,31 +41,32 @@ var (
 
-func noop(ctx *log.Context, h HandlerEnvironment, seqNum int) (string, error) {
-	ctx.Log("event", "noop")
+func noop(ctx logging.Logger, h handlerenv.HandlerEnvironment, seqNum int) (string, error) {
```
this pattern of calling the parameter `ctx` seems a bit confusing, why not rename it to `logger`?

klugorosado replied:
ack
